core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -783,6 +817,38 @@ private[spark] class TaskSetManager(
             s"Lost task ${info.id} in stage ${taskSet.id} (TID $tid) on ${info.host}, executor" +
               s" ${info.executorId}: ${ef.className} (${ef.description}) [duplicate $dupCount]")
         }
+
+        // for next round increase cpusPerTask for OOME/LME
+        if (supportsDynamicCpusPerTask && !isZombie && (ef.className.contains("OutOfMemory") ||
+            ef.className.contains("LowMemoryException"))) {
+          hasDynamicCpusPerTask = true
+          val task = tasks(index)
+          // apply a reasonable upper limit on dynamic cpusPerTask
+          if (task.cpusPerTask < min(confCpusPerTask + 4, sched.maxAvailableCpus / 2)) {
+            task.cpusPerTask += 1
+            // update maxCpusPerTask tracked in the TaskSetManager which is
+            // required for the check in TaskSchedulerImpl.resourceOfferSingleTaskSet
+            if (task.cpusPerTask > maxCpusPerTask) {
```

Comment:
Multiple handleFail tasks can happen simultaneously. Do you see any race on updating maxCpusPerTask here, which is a shared number across TaskSet?

Reply:
All calls in TaskSetManager are guarded by sync block at the TaskSetSchedulerImpl level.